### PR TITLE
MRG: suggested fixes for header check errors

### PR DESCRIPTION
**Pull request by matthew-brett**
Changing the header checks in the Cifti2 header.

Updating the Cifti2 header before saving, to avoid warnings from header checks
by Nifti2 code.

### CODE REVIEW

File: nibabel/cifti2/parse_cifti2_fast.py [hunk 1]
```diff
@@ -343,7 +376,7 @@ def StartElementHandler(self, name, attrs):
             model = Cifti2BrainModel()
             mim = self.struct_state[-1]
             if not isinstance(mim, Cifti2MatrixIndicesMap):
-                raise CIFTI2HeaderErrorCIFTI2HeaderError(
```
Comment (matthew-brett) [inline nibabel/cifti2/parse_cifti2_fast.py:346]: I guess this guy wasn't being tested?

File: nibabel/cifti2/parse_cifti2_fast.py [hunk 2]
```diff
@@ -78,12 +79,43 @@ class _Cifti2AsNiftiHeader(Nifti2Header):
     ''' Class for Cifti2 header extension '''
 
     @classmethod
+    def _valid_intent_code(klass, intent_code):
+        """ Return True if `intent_code` matches our class `klass`
+        """
+        return intent_code >= 3000 and intent_code < 3100  # and intent_code != 3002
```
Comment (matthew-brett) [inline nibabel/cifti2/parse_cifti2_fast.py:85]: Should `intent_code != 3002` be uncommented?

File: nibabel/cifti2/parse_cifti2_fast.py (hunk 1 above)
Comment (demianw) [inline nibabel/cifti2/parse_cifti2_fast.py:346]: Yes there is still a lack of testing for badly formed headers. However, I think the overall testing coverage is pretty good.

File: nibabel/cifti2/parse_cifti2_fast.py (hunk 2 above)
Comment (demianw) [inline nibabel/cifti2/parse_cifti2_fast.py:85]: Actually, it should be deleted. In this way Cifti2Image handles all CIFTI2 cases. I think that generating the classes for each image subtype, such as Cifti2DenseDataSeries, should be either aborted or left for a future PR.

If not it's generating all of the or none. I think adding them later will not change the current interface but extend it. Hence, not a big problem in leaving it for later.

**Pull request merged**